Plugins/Public/base_plugin/FactoryModule.cpp

```diff
@@ -90,181 +125,196 @@ wstring FactoryModule::GetInfo(bool xml)
 bool FactoryModule::Timer(uint time)
 {
 
-	if ((time%set_tick_time) != 0)
+	if ((time % set_tick_time) != 0)
+	{
 		return false;
+	}
 
 	// Get the next item to make from the build queue.
-	if (!active_recipe.nickname && build_queue.size())
+	if (!active_recipe.nickname && !build_queue.empty())
 	{
-		map<uint, RECIPE>::iterator i = recipes.find(build_queue.front());
-		if (i != recipes.end())
-		{
-			active_recipe = i->second;
-		}
+		SetActiveRecipe(build_queue.front());
 		build_queue.pop_front();
 	}
 
 	// Nothing to do.
-	if (!active_recipe.nickname)
-		return false;
-
-	if (Paused)
+	if (!active_recipe.nickname || !base->isCrewSupplied || Paused)
+	{
 		return false;
+	}
 
 	// Consume goods at the cooking rate.
 	bool cooked = true;
-	for (map<uint, uint>::iterator i = active_recipe.consumed_items.begin();
-		i != active_recipe.consumed_items.end(); ++i)
+
+	for (const auto& catalyst : active_recipe.catalyst_items)
+	{
+		uint good = catalyst.first;
+		uint quantityNeeded = catalyst.second;
+
+		uint presentAmount = base->HasMarketItem(good);
+		if ((presentAmount - base->reservedCatalystMap[good]) < quantityNeeded)
+		{
+			return false;
+		}
+		base->reservedCatalystMap[good] += quantityNeeded;
+	}
+
+	if (active_recipe.credit_cost)
 	{
-		uint good = i->first;
-		uint quantity = i->second > active_recipe.cooking_rate ? active_recipe.cooking_rate : i->second;
-		if (quantity)
+		uint moneyToRemove = min(active_recipe.cooking_rate * 10, active_recipe.credit_cost);
+		if (base->money >= moneyToRemove)
+		{
+			base->money -= moneyToRemove;
+			active_recipe.credit_cost -= moneyToRemove;
+		}
+		if (active_recipe.credit_cost)
 		{
 			cooked = false;
-			map<uint, MARKET_ITEM>::iterator market_item = base->market_items.find(good);
-			if (market_item != base->market_items.end())
-			{
-				if (market_item->second.quantity >= quantity)
-				{
-					i->second -= quantity;
-					base->RemoveMarketGood(good, quantity);
-					return false;
-				}
-			}
+		}
+	}
+
+	for (auto& i : active_recipe.consumed_items)
+	{
+		uint good = i.first;
+		uint quantity = min(active_recipe.cooking_rate, i.second);
```

Comment:
is this min supposed to be here?

Reply:
Yes, If for whatever reason the remaining commodities end up being lower than the cooking rate, we only want to consume the smaller of the two, so a recipe requiring 500 of Silvery_Metal and a cooking rate of 200 will consume 200+200+100 instead of 200+200+200